### ISSUE TITLE
Segmentation Fault on M1 MacOS

### ISSUE THREAD

**Issue by TimJSwan89** [xmrig/xmrig#3183]
Built using the recommended commands at https://xmrig.com/docs/miner/build/macos
```
brew install cmake libuv openssl hwloc
git clone https://github.com/xmrig/xmrig.git
mkdir xmrig/build && cd xmrig/build
cmake .. -DUV_INCLUDE_DIR=/opt/homebrew/include -DUV_LIBRARY=/opt/homebrew/lib/libuv.a -DOPENSSL_ROOT_DIR=/opt/homebrew/opt/openssl -DHWLOC_INCLUDE_DIR=/opt/homebrew/include -DHWLOC_LIBRARY=/opt/homebrew/lib/libhwloc.dylib
make -j$(sysctl -n hw.logicalcpu)
```
```
./xmrig --donate-level 10 --opencl -o stratum+tcp://stratum.cudopool.com:30010 -u o:744943:n:MBA-Tim -p x -k --nicehash --coin monero -a rx/0
 * ABOUT        XMRig/6.18.1 clang/14.0.0
 * LIBS         libuv/1.44.2 OpenSSL/3.0.7 hwloc/2.8.0
 * HUGE PAGES   unavailable
 * 1GB PAGES    unavailable
 * CPU          Apple M1 (1) 64-bit AES
                L2:16.0 MB L3:0.0 MB 8C/8T NUMA:1
 * MEMORY       7.7/8.0 GB (96%)
 * DONATE       10%
 * POOL #1      stratum+tcp://stratum.cudopool.com:30010 coin Monero
 * COMMANDS     hashrate, pause, resume, results, connection
 * OPENCL       #0 Apple/OpenCL 1.2 (Sep 30 2022 01:38:14)
 * OPENCL GPU   #0 n/a Apple M1 1000 MHz cu:8 mem:1024/5461 MB
 * CUDA         disabled
[2022-12-22 21:56:33.151]  net      use pool stratum.cudopool.com:30010  147.135.37.197
[2022-12-22 21:56:33.152]  net      new job from stratum.cudopool.com:30010 diff 45000 algo rx/0 height 2783311 (13 tx)
[2022-12-22 21:56:33.153]  cpu      use argon2 implementation default
[2022-12-22 21:56:33.153]  randomx  init dataset algo rx/0 (8 threads) seed fd64b331392849ec...
[2022-12-22 21:56:33.154]  randomx  allocated 2336 MB (2080+256) huge pages 0% 0/1168 +JIT (1 ms)
[2022-12-22 21:56:39.590]  randomx  dataset ready (6436 ms)
[2022-12-22 21:56:39.590]  cpu      use profile  rx  (8 threads) scratchpad 2048 KB
[2022-12-22 21:56:39.591]  cpu      READY threads 8/8 (8) huge pages 0% 0/8 memory 16384 KB (1 ms)
zsh: segmentation fault  ./xmrig --donate-level 10 --opencl -o stratum+tcp://stratum.cudopool.com:3001
```
```
otool -L xmrig                   
xmrig:
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1228.0.0)
	/opt/homebrew/opt/hwloc/lib/libhwloc.15.dylib (compatibility version 22.0.0, current version 22.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.36.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1953.255.0)
```
If you need the commit hash, in the build folder:
```
git rev-parse HEAD
28e81bd7c09c88f8dfe5ab5fb1fbd62f3bca1f8b
```

**Comment by TimJSwan89** [comment]
Scratch the OpenCL. The Seg fault happens without it.
```
./xmrig --donate-level 10 -o stratum+tcp://stratum.cudopool.com:30010 -u o:744943:n:MBA-Tim -p x -k --nicehash --coin monero -a rx/0
```
The only reason why I thought it depended on it was because earlier I had it running CPU only with another build, which I think I had simply downloaded from the website, which failed with an OpenCL metal dependency.

**Comment by Spudz76** [comment]
Use `./scripts/build_deps.sh` for the openssl/hwloc/libuv dependencies, the homebrew ones almost never work.

**Comment by Unnameless** [comment]
> Use `./scripts/build_deps.sh` for the openssl/hwloc/libuv dependencies, the homebrew ones almost never work.

Actually most of the type brew installed dependencies always worked for me, in the last couple of years, ever since xmrig was on CN. This time something went sideways.

**Comment by Spudz76** [comment]
> the homebrew ones almost never work

in other words, bundled versions are the tested ones so if you like support, you use the bundled ones.

**Comment by Unnameless** [comment]
> Use `./scripts/build_deps.sh` for the openssl/hwloc/libuv dependencies, the homebrew ones almost never work.

actually seg fault happens when compiling using the script also! The only way to avoid it, is to use the already compiled binaries

**Comment by furic** [comment]
Ran `./scripts/build_deps.sh` and still having `zsh: segmentation fault` on the build.
I'm building 6.19.0, in macOS 13.2.1.

**Comment by giffeler** [comment]
xmrig 6.19.2, macOS 13.3. Independent of the chosen compiler (gcc or current xcode), independent of homebrew: segfault.

**Comment by MarcusNewman** [comment]
Still seeing this issue, M2 XMRig 6.19.3
Using build_deps.sh shows: * LIBS         libuv/1.45.0 OpenSSL/3.1.1 hwloc/2.9.0
precompiled binary shoes: * LIBS         libuv/1.44.2 OpenSSL/3.0.7 hwloc/2.9.0
Another thought, if the build steps at https://xmrig.com/docs/miner/build/macos are incorrect, how can we the users update them?

**Comment by SChernykh** [comment]
This is a known problem, and the only solution so far is to use older MacOS SDK for the build: https://github.com/tevador/RandomX/issues/262#issuecomment-1509895418
So try to add `-DCMAKE_OSX_SYSROOT=~/dev/MacOSX12.3.sdk` (download the SDK first of course) to cmake command line and do a fresh build.

**Comment by katletkis** [comment]
Thank you @SChernykh, building with OSX12.3.sdk worked for me

**Comment by alicedb2** [comment]
I'm compiling on a M2 Macbook Air running Ventura 13.4.2. Using `-DCMAKE_OSX_SYSROOT=/path/to/MacOSX12.3.sdk` when running cmake fixed the segfault for me as well, but the annoying thing is that during linking I did get a bunch of warnings
```
ld: warning: object file (/Users/alice/src/xmrig-md-6.18.1/scripts/deps/lib/libhwloc.a(topology.o)) was built for newer macOS version (13.0) than being linked (12.3)
ld: warning: object file (/Users/alice/src/xmrig-md-6.18.1/scripts/deps/lib/libuv.a(libuv_la-fs-poll.o)) was built for newer macOS version (13.0) than being linked (12.3)
ld: warning: object file (/Users/alice/src/xmrig-md-6.18.1/scripts/deps/lib/libssl.a(bio_ssl.o)) was built for newer macOS version (13.0) than being linked (12.3)
[...]
```
Obviously the 13.3 SDK was picked up by libuv/libhwloc/libssl during `build_deps.sh`.

What ended up working cleanly was to set the `SDKROOT` environmental prior to both `build_deps.sh` and `cmake`, so something like

```
export SDKROOT=Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk
./build_deps.sh
cd ../build
cmake .. -DOPENSSL_CRYPTO_LIBRARY=../scripts/deps/lib/libcrypto.a -DOPENSSL_SSL_LIBRARY=../scripts/deps/lib/libssl.a -DOPENSSL_INCLUDE_DIR=../scripts/deps/include -DHWLOC_INCLUDE_DIR=../scripts/deps/include -DHWLOC_LIBRARY=../scripts/deps/lib/libhwloc.a
make
```
No more segfaults yes, but also no more warnings against mismatched versions during linking.

**Comment by SChernykh** [comment]
Fixed in #3346
